src/site/_filters/filter-by-tag.js

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Returns posts filtering by specific tags.
+ */
+
+const {findByUrl} = require('./find-by-url');
+
+/**
+ * @param {EleventyCollectionObject} _ An eleventy posts object.
+ * @param {string[]} tags Array of tag name.
+ * @return {EleventyCollectionItem} Latest post filter by tag.
+ */
+
+function latestPostByTags(_, tags) {
+  let posts = [];
+  tags.forEach((tag) => {
+    // @ts-ignore
+    posts = [...posts, this.ctx.collections[tag].pop()];
+  });
```

Comment:
These are wild three lines 😅  I think what you are looking for is `Array.prototype.map`, since you are creating an array from another array.

L33 is problematic in two ways: you are destructuring an array (`...posts`) just to add one element, while you could instead add to the existing array (`Array.prototype.push`). But the other problematic bit is `.pop()`, as it alters the original array, which we do not own in this context - so that's evil.

I think this can be rewritten as:

```
const posts = tags.map((tag) => {
  const tagPosts = this.ctx.collections[tag];
  return tagPosts[tagPosts.length - 1];
}
```

But you can actually improve this even more and get rid of the sorting, if you use an intermediate variable to store current newest post, like this:

```js
let newestPost = null;
for (const tag of tags) {
  const tagPosts = this.ctx.collections[tag];
  const candidatePost = tagPosts[tagPosts.length - 1];

  // Accounts for the first run of the loop
  if (!newestPost) { 
    newestPost = candidatePost; 
    return;
  }

  // Check if the current candidate post is newer then the newest
  // post that is currently stored
  if (new Date(newestPost.date) > new Date(candidatePost.date)) {
    newestPost = candidatePost;
  }
}
```

That's slightly more complicated to read, but does the job in one loop (while `map` and `sort` would be two). Basically a fun interview question...

Reply:
I have refactored this logic. Thank you for your suggestion.